### PR TITLE
Stop failing acceptance tests if their directories already exist

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

After PR #1560, the cached sapling tests started failing, because the acceptance test utility functions were too strict about existing directories.

## Solution

Fixes:
  * ignore existing directories
  * consistently create missing config and state directories as needed
  * only ignore NotFound errors when deleting the previous config in replace_config

Cleanup:
  * refactor the common config writing code into a separate function

The code in this pull request has:
  - [x] Documentation Comments
  - [x] Passes Existing Unit Tests
  - We won't know if the cached sapling tests pass until we merge this PR

## Review

@dconnolly knows the most about this code, so she should review it eventually.

This bug is causing failures on main, so it's pretty urgent, I hope @yaahc can do a quick review.

## Related Issues

Makes it harder to test sync issues like #1586

### CODE REVIEW

File: zebrad/tests/acceptance.rs [hunk 1]
```diff
@@ -68,15 +68,30 @@ where
...
-    /// Overwrite any existing config the test directory and use it for all
+    /// Overwrite any existing config, and use the newly written config for all
     /// subsequently spawned processes.
+    ///
+    /// Recursively create directories for the config and state as needed.
     fn replace_config(self, config: ZebradConfig) -> Result<Self>;
+
+    /// Config writing helper for [`with_config`] and [`replace_config`].
+    ///
+    /// If needed:
```
Comment (dconnolly) [inline zebrad/tests/acceptance.rs:101]: 👍

**Comment by dconnolly** [comment]
Duplicated this branch into the zebra base repo and manually triggered the Test workflow on it: https://github.com/ZcashFoundation/zebra/actions/runs/484232685

**Review by yaahc — DISMISSED** [review]

**Comment by teor2345** [comment]
> Duplicated this branch into the zebra base repo and manually triggered the Test workflow on it: https://github.com/ZcashFoundation/zebra/actions/runs/484232685

Oops, this new code doesn't work, because it overwrites the configured state directory, adding a "state" directory to the provided path.

I'll go rewrite the utility functions so we have a "just write the config and don't mess with it" function.

**Comment by teor2345** [comment]
We also had a testnet failure, that's not surprising, we'll fix that by deploying more testnet nodes.

File: zebrad/tests/acceptance.rs [hunk 2]
```diff
@@ -784,7 +826,7 @@ fn create_cached_database_height(network: Network, height: Height) -> Result<()>
 
     let dir = PathBuf::from("/zebrad-cache");
     let mut child = dir
-        .with_config(config)?
+        .with_exact_config(&config)?
```
Comment (teor2345) [inline zebrad/tests/acceptance.rs:829]: @dconnolly @yaahc this is the actual fix, all the rest is refactoring and type system enforcement

File: zebrad/tests/acceptance.rs [hunk 3]
```diff
@@ -103,47 +135,57 @@ where
         }
     }
 
-    fn with_config(self, mut config: ZebradConfig) -> Result<Self> {
+    fn with_config(self, config: &mut ZebradConfig) -> Result<Self> {
+        self.cache_config_update_helper(config)?
+            .write_config_helper(config)
+    }
+
+    fn with_exact_config(self, config: &ZebradConfig) -> Result<Self> {
```
Comment (teor2345) [inline zebrad/tests/acceptance.rs:143]: This is the new utility function, all the rest is refactoring and type system enforcement

Comment (teor2345) [reply zebrad/tests/acceptance.rs:143]: Oh and better documentation

**Comment by teor2345** [comment]
@dconnolly can you do another manual run?

**Comment by teor2345** [comment]
Hmm yeah looks like testnet is a bit sad right now

> Jan 14 13:25:43.604  WARN {zebrad="92d95d4b" net="Test"}:sync: zebrad::components::sync: error downloading and verifying block e=Elapsed(())

**Review by dconnolly — APPROVED** [review]

**Comment by dconnolly** [comment]
Looks good! 

https://github.com/ZcashFoundation/zebra/runs/1699955240?check_suite_focus=true

**Pull request merged**